[SPARK-17219][ML] enhanced NaN value handling in Bucketizer

VinceShieh opened the pull request:
## What changes were proposed in this pull request?

This PR is an enhancement of PR with commit ID:57dc326bd00cf0a49da971e9c573c48ae28acaa2.
NaN is a special type of value which is commonly seen as invalid. But We find that there are certain cases where NaN are also valuable, thus need special handling. We provided user when dealing NaN values with 3 options, to either reserve an extra bucket for NaN values, or remove the NaN values, or report an error, by setting handleNaN "keep", "skip", or "error"(default) respectively.

'''Before:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
'''After:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
          .setHandleNaN("keep")
## How was this patch tested?

Tests added in QuantileDiscretizerSuite, BucketizerSuite and DataFrameStatSuite

Signed-off-by: VinceShieh vincent.xie@intel.com

Commit:
[SPARK-17219][ML] enchance NaN value handling in Bucketizer

This PR is an enhancement of PR with commit ID:57dc326bd00cf0a49da971e9c573c48ae28acaa2.
We provided user when dealing NaN value in the dataset with 3 options, to either reserve an extra
bucket for NaN values, or remove the NaN values, or report an error, by setting "keep", "skip",
or "error"(default) to handleInvalid.

'''Before:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
'''After:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
          .setHandleNaN("skip")

Signed-off-by: VinceShieh <vincent.xie@intel.com>

SparkQA commented:
**[Test build #66731 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66731/consoleFull)** for PR 15428 at commit [`a3e4308`](https://github.com/apache/spark/commit/a3e43086dcf6ecee20461567e2cc506db29f80a7).

SparkQA commented:
**[Test build #66731 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66731/consoleFull)** for PR 15428 at commit [`a3e4308`](https://github.com/apache/spark/commit/a3e43086dcf6ecee20461567e2cc506db29f80a7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66731/
Test FAILed.

SparkQA commented:
**[Test build #66733 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66733/consoleFull)** for PR 15428 at commit [`cd8113c`](https://github.com/apache/spark/commit/cd8113c456870dd89754d0d48bfe6d0931ad05bb).

srowen commented on mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala:
```diff
@@ -270,10 +270,10 @@ private[ml] trait HasFitIntercept extends Params {
 
   /**
-   * Param for how to handle invalid entries. Options are skip (which will filter out rows with bad values), or error (which will throw an error). More options may be added later.
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with bad values), or error (which will throw an error), or keep (which will keep the bad values in certain way). More options may be added later.
```
I'm neutral on the complexity that this adds, but not against it. It gets a little funny to say "keep invalid data" but I think we discussed that on the JIRA

VinceShieh replied:
yeah, usually we treat NaN as a type of invalid value, but we know there are cases they are proved useful, so, it's a bit dilemma here, also with the API naming.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+    val bucketizer: UserDefinedFunction = udf { (feature: Double, flag: String) =>
+      Bucketizer.binarySearchForBuckets($(splits), feature, flag)
+    }
+    val filteredDataset = {
```
Doesn't this need to try to handle "error"?

```
val filteredDataSet = getHandleInvalid match {
  case "skip" => dataset.na.drop
  case "keep" => dataset
  case "error" => 
    if (...dataset contains NaN...) {
      throw new IllegalArgumentException(...)
   } else {
      dataset
   }
}
```

srowen replied:
I don't see that the method handles `NaN` below. What binarySearch returns is undefined. One place or the other I think this has to be explicitly handled.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -128,8 +145,9 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
-  private[feature] def binarySearchForBuckets(splits: Array[Double], feature: Double): Double = {
-    if (feature.isNaN) {
+  private[feature] def binarySearchForBuckets
+  (splits: Array[Double], feature: Double, flag: String): Double = {
```
Nit: I think the convention is to leave the open paren on the previous line

Doesn't this need to handle "skip" and "error"? throw an exception on NaN if "error" or ignore it if "skip"?

SparkQA commented:
**[Test build #66735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66735/consoleFull)** for PR 15428 at commit [`5cd58b7`](https://github.com/apache/spark/commit/5cd58b776883d5185f2b050317d260b924ebef5e).

SparkQA commented:
**[Test build #66733 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66733/consoleFull)** for PR 15428 at commit [`cd8113c`](https://github.com/apache/spark/commit/cd8113c456870dd89754d0d48bfe6d0931ad05bb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66733/
Test PASSed.

SparkQA commented:
**[Test build #66735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66735/consoleFull)** for PR 15428 at commit [`5cd58b7`](https://github.com/apache/spark/commit/5cd58b776883d5185f2b050317d260b924ebef5e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66735/
Test PASSed.

srowen reviewed (changes requested):
I'll let @thunterdb or @jkbradley mostly review from here to see if this is what they intended

jkbradley commented:
Thanks! I'll take a look.

Could you please fix the typo in the title? "enchanced" -> "enhanced"

VinceShieh commented:
typo corrected. Thank you all. @srowen @jkbradley

jkbradley commented on docs/ml-features.md:
```diff
@@ -1104,9 +1104,11 @@ for more details on the API.
-distinct values of the input to create enough distinct quantiles. Note also that NaN values are
-handled specially and placed into their own bucket. For example, if 4 buckets are used, then
-non-NaN data will be put into buckets[0-3], but NaNs will be counted in a special bucket[4].
+distinct values of the input to create enough distinct quantiles. Note also that QuantileDiscretizer
```
(same as below) Is "possible that the number of buckets used will be less than this value" true?  It was true before this used Dataset.approxQuantiles, but I don't think it is true any longer.

VinceShieh replied:
In cases when the number of buckets requested by the user is greater than the number of distinct splits generated from Bucketizer, the returned number of buckets will be less than requested.

jkbradley replied:
Yep, you're right

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
     transformSchema(dataset.schema)
-    val bucketizer = udf { feature: Double =>
-      Bucketizer.binarySearchForBuckets($(splits), feature)
+    val bucketizer: UserDefinedFunction = udf { (feature: Double, flag: String) =>
```
Also, at this point, you already know the value for flag, so you can just use it here, rather than making the UDF take an extra argument.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+      if ("skip" == getHandleInvalid) {
+        dataset.na.drop
+      } else {
+        dataset
```
I'd put ```dataset.toDF()``` here and not import existentials.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -128,8 +145,11 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
+  private[feature] def binarySearchForBuckets(
+                                               splits: Array[Double],
+                                               feature: Double,
+                                               flag: String): Double = {
```
Use a better name than "flag"  How about "keepNaNs"?
Also, please document this argument.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
- * NaN values are handled specially and placed into their own bucket. For example, if 4 buckets
- * are used, then non-NaN data will be put into buckets(0-3), but NaNs will be counted in a special
- * bucket(4).
+ * possible that the number of buckets used will be less than this value, for example, if there are
```
Is this true?  It was true before this used Dataset.approxQuantiles, but I don't think it is true any longer.

VinceShieh replied:
same as the comment above

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
+ * possible that the number of buckets used will be less than this value, for example, if there are
+ * too few distinct values of the input to create enough distinct quantiles. Note also that
+ * QuantileDiscretizer will raise an error when it finds NaN value in the dataset, but user can
+ * also choose to either keep or remove NaN values within the dataset by calling setHandleInvalid.
```
"calling setHandleInvalid" -> "setting `handleInvalid`"

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
+ * too few distinct values of the input to create enough distinct quantiles. Note also that
+ * QuantileDiscretizer will raise an error when it finds NaN value in the dataset, but user can
+ * also choose to either keep or remove NaN values within the dataset by calling setHandleInvalid.
+ * If user chooses to keep NaN values, they will be handled specially and placed into their own
```
This documentation should go in the handleInvalid Param doc string.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala:
```diff
@@ -114,6 +115,7 @@ class BucketizerSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
           .setInputCol("feature")
```
Just set splits here; there's no need to set the other Params for this test.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala:
```diff
@@ -83,10 +83,20 @@ class QuantileDiscretizerSuite
-    val observedNumBuckets = result.select("result").distinct.count
+    // Reserve extra one bucket for NaN values
+    discretizer.setHandleInvalid("keep")
+    var expectedNumBuckets = discretizer.fit(df).getSplits.length - 1
```
I'd also recommend directly testing the values of the splits.  The current test makes sure that handleInvalid is passed to the bucketizer correctly, which is important but separate.

Also, please use vals (not vars) for clarity.  I'd recommend making a helper method for lines 87-93, which can be reused for the test of handleInvalid = "skip"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala:
```diff
@@ -83,10 +83,20 @@ class QuantileDiscretizerSuite
+
+    // Remove NaN values
+    discretizer.setHandleInvalid("skip")
+    expectedNumBuckets = discretizer.fit(df).getSplits.length - 2
```
Same here; I'd also recommend directly testing the values of the splits.

jkbradley commented on python/pyspark/ml/feature.py:
```diff
@@ -1157,9 +1157,11 @@ class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadab
     categorical features. The number of bins can be set using the :py:attr:`numBuckets` parameter.
     It is possible that the number of buckets used will be less than this value, for example, if
```
Here too: no longer the case

VinceShieh replied:
same as the comment above.

jkbradley reviewed (commented):
Done with review.  Thanks for the PR!

Other comments:
- Add unit test for `DataFrameStatFunctions.approxQuantile` to check for handling of NaN values.
- Create JIRA for adding handleInvalid to Python API.

SparkQA commented:
**[Test build #67231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67231/consoleFull)** for PR 15428 at commit [`c350e9f`](https://github.com/apache/spark/commit/c350e9fd0d20ad7114644d53df83231548171f89).

SparkQA commented:
**[Test build #67231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67231/consoleFull)** for PR 15428 at commit [`c350e9f`](https://github.com/apache/spark/commit/c350e9fd0d20ad7114644d53df83231548171f89).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67231/
Test FAILed.

SparkQA commented:
**[Test build #67232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67232/consoleFull)** for PR 15428 at commit [`0fb8d38`](https://github.com/apache/spark/commit/0fb8d38dca953b2fb1f05b8debb263ed26e0ed36).

VinceShieh commented:
Thanks for your valuable suggestions. @jkbradley @srowen

SparkQA commented:
**[Test build #67233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67233/consoleFull)** for PR 15428 at commit [`d1dd840`](https://github.com/apache/spark/commit/d1dd84035802f5aad620aae2953bed4842b49aa8).

SparkQA commented:
**[Test build #67232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67232/consoleFull)** for PR 15428 at commit [`0fb8d38`](https://github.com/apache/spark/commit/0fb8d38dca953b2fb1f05b8debb263ed26e0ed36).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67232/
Test PASSed.

SparkQA commented:
**[Test build #67233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67233/consoleFull)** for PR 15428 at commit [`d1dd840`](https://github.com/apache/spark/commit/d1dd84035802f5aad620aae2953bed4842b49aa8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67233/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
-    val bucketizer = udf { feature: Double =>
-      Bucketizer.binarySearchForBuckets($(splits), feature)
+
+    val bucketizer: UserDefinedFunction = udf { (feature: Double) =>
```
Ah, sorry, one more comment.  I'm not quite sure how closure capture behaves currently, but it might be good to define local vals for ```$(splits)``` and ```getHandleNaN.isDefined && getHandleNaN.get```.  Since these reference methods in the Bucketizer class, I believe the UDF may capture the whole Bucketizer class instead of just those vals.

After you define them in local vals here, you can use those vals in this UDF.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala:
```diff
@@ -82,14 +82,23 @@ class QuantileDiscretizerSuite
+
+    for (handleNaN <- Array("keep", "skip")) {
+      discretizer.setHandleNaN(handleNaN)
+      val expectedNumBuckets = {
```
This is kind of complex logic just to get the expected number of splits.  I'd recommend just putting the expected bucket *values* in the original DataFrame:
```
val df = sc.parallelize(Seq(
  (1.0, /*expected value for option "keep"*/, /*expected value for option "skip"*/),
  ...
)).toDF("input", "expectedKeep", "expectedSkip")
```
Then you can compare with the actual values.  That'll be an easier test to understand IMO.

jkbradley commented on python/pyspark/ml/feature.py:
```diff
@@ -1157,9 +1157,11 @@ class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadab
-    that NaN values are handled specially and placed into their own bucket. For example, if 4
-    buckets are used, then non-NaN data will be put into buckets(0-3), but NaNs will be counted in
-    a special bucket(4).
+    that QuantileDiscretizer will raise an error when it finds NaN value in the dataset, but user
```
Actually no need to update Python API until it is updated to include handleNaN

jkbradley replied:
This isn't available in Python yet, so can you please revert this change to feature.py?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -100,6 +102,24 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
+   * @group param
+   */
+  @Since("2.1.0")
+  val handleNaN: Param[String] = new Param[String](this, "handleNaNs", "how to handle NaN" +
```
Add getHandleNaN

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -100,6 +102,24 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
+   * Param for how to handle NaN entries. Options are skip (which will filter out rows with
+   * NaN values), or error (which will throw an error), or keep (which will make the NaN
+   * values an extra bucket). More options may be added later.
+   *
```
State default

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+  /**
+   * Param for how to handle NaN entries. Options are skip (which will filter out rows with
+   * NaN values), or error (which will throw an error), or keep (which will make the NaN
+   * values an extra bucket). More options may be added later.
```
Remove "More options may be added later."

Also state default.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+  val handleNaN: Param[String] = new Param[String](this, "handleNaNs", "how to handle NaN" +
+    "entries. Options are skip (which will filter out rows with NaN values), or error" +
+    "(which will throw an error), or keep (which will make the NaN values an extra bucket)." +
+    "More options may be added later", ParamValidators.inArray(Array("skip", "error", "keep")))
```
Remove "More options may be added later."

jkbradley reviewed (commented):
Thanks for the updates!  A few more comments.

Also, I see you renamed the param to handleNaN.  Do you think we will ever want to handle null values too?  It might be good to stick with "handleInvalid" in case we add null support.

SparkQA commented:
**[Test build #67386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67386/consoleFull)** for PR 15428 at commit [`47aad24`](https://github.com/apache/spark/commit/47aad243f2d1dda73338fc234283ea520e1f848e).

SparkQA commented:
**[Test build #67386 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67386/consoleFull)** for PR 15428 at commit [`47aad24`](https://github.com/apache/spark/commit/47aad243f2d1dda73338fc234283ea520e1f848e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67386/
Test PASSed.

SparkQA commented:
**[Test build #67429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67429/consoleFull)** for PR 15428 at commit [`70cee57`](https://github.com/apache/spark/commit/70cee5791d09f596e55f39da1c4d0386a1b7e7a3).

SparkQA commented:
**[Test build #67430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67430/consoleFull)** for PR 15428 at commit [`b14fbab`](https://github.com/apache/spark/commit/b14fbab7487a8464ba2a53bb9804e00fd14d3785).

SparkQA commented:
**[Test build #67430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67430/consoleFull)** for PR 15428 at commit [`b14fbab`](https://github.com/apache/spark/commit/b14fbab7487a8464ba2a53bb9804e00fd14d3785).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #67429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67429/consoleFull)** for PR 15428 at commit [`70cee57`](https://github.com/apache/spark/commit/70cee5791d09f596e55f39da1c4d0386a1b7e7a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67430/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67429/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +74,52 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+  /**
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with
+   * invalid values), or error (which will throw an error), or keep (which will keep the invalid
+   * values in certain way). Default behaviour is to report an error for invalid entries.
```
I'd just write: ```default: "error"```
Rewording as "report" instead of "throw" could confuse people.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:
```diff
@@ -73,15 +74,52 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
+
+  /** @group getParam */
+  @Since("2.1.0")
+  def gethandleInvalid: Option[Boolean] = $(handleInvalid) match {
```
This should just return ```$(handleInvalid)```, just like any other Param getter method.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala:
```diff
@@ -100,6 +102,33 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
+  /**
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with
+   * invalid values), or error (which will throw an error), or keep (which will keep the invalid
+   * values in certain way). Default behaviour is to report an error for invalid entries.
```
I'd just write: ```default: "error"```
Rewording as "report" instead of "throw" could confuse people.

jkbradley reviewed (commented):
Thanks for the updates.
Also, when you push further updates, could you please just push a new commit?  Rebasing makes it harder to review since it makes it impossible to connect your most recent changes with the recent reviewer comments.  Thanks!

Commit:
[SPARK-17219][ML] enchance NaN value handling in Bucketizer

This PR is an enhancement of PR with commit ID:57dc326bd00cf0a49da971e9c573c48ae28acaa2.
We provided user when dealing NaN value in the dataset with 3 options, to either reserve an extra
bucket for NaN values, or remove the NaN values, or report an error, by setting "keep", "skip",
or "error"(default) to handleInvalid.

'''Before:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
'''After:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
          .setHandleNaN("skip")

Signed-off-by: VinceShieh <vincent.xie@intel.com>

SparkQA commented:
**[Test build #67493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67493/consoleFull)** for PR 15428 at commit [`5274d4a`](https://github.com/apache/spark/commit/5274d4a3703193a59607635f80eb9e3ebe61552c).

SparkQA commented:
**[Test build #67493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67493/consoleFull)** for PR 15428 at commit [`5274d4a`](https://github.com/apache/spark/commit/5274d4a3703193a59607635f80eb9e3ebe61552c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67493/
Test PASSed.

jkbradley commented:
Thanks, you still need to remove the change in feature.py, but other than that, this should be ready.

Commit:
revert changes in feature.py

Signed-off-by: VinceShieh <vincent.xie@intel.com>

VinceShieh commented:
sorry, I must have forgotten to commit the changes.
All done now. Thanks for reviewing.

SparkQA commented:
**[Test build #67547 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67547/consoleFull)** for PR 15428 at commit [`2f98d31`](https://github.com/apache/spark/commit/2f98d31118413e61e1aa0431da402c41aa1ca5a6).

SparkQA commented:
**[Test build #67547 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67547/consoleFull)** for PR 15428 at commit [`2f98d31`](https://github.com/apache/spark/commit/2f98d31118413e61e1aa0431da402c41aa1ca5a6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67547/
Test PASSed.

jkbradley committed:
Cleanups: docs cleanups, slightly improved unit test coverage, fixed naming of set/get for handleInvalid

jkbradley commented:
Thanks for the update!  On a last glance through, I spotted a few things to fix up.  Since some had to do with wording, I thought it'd be easier just to send a PR to your PR.  Can you please review this and merge it if it looks Ok to you?  Thanks!  Here it is: [https://github.com/VinceShieh/spark/pull/2]

Commit:
Merge pull request #2 from jkbradley/VinceShieh-spark-17219_followup

Small cleanups

SparkQA commented:
**[Test build #67611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67611/consoleFull)** for PR 15428 at commit [`2b1b81d`](https://github.com/apache/spark/commit/2b1b81d43272403cad524236a69d9ab90c752017).

SparkQA commented:
**[Test build #67611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67611/consoleFull)** for PR 15428 at commit [`2b1b81d`](https://github.com/apache/spark/commit/2b1b81d43272403cad524236a69d9ab90c752017).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67611/
Test PASSed.

jkbradley commented:
Thanks for merging that!  This LGTM
Merging with master
Thanks a lot @VinceShieh !
